### DISCUSSION TITLE
Reference frames, loadings, and deflections

### DISCUSSION THREAD

**Discussion by Cardoza2** [OpenFAST/openfast#1494] [Q&A]
Hi everyone,

TLDR: I have negative deflections in the Y direction from a positive Y loading and I can't figure out why. Any ideas? 

I've run into some behavior that I don't quite understand. I think it is probably user error. I'm simulating a simplified version of the NREL 5MW wind turbine. I made the following modifications:
- Set the sweep, precone, curve, pitch, aerodynamic and structural twists to zero. 
- I made the chord and airfoil are constant. 
- Left the stiffness and mass matrix the same, but I did linearly interpolate them so I could have the properties defined at arbitrary points.
- Turned off gravity, aerodynamic moments, and all of the DOF expect for the flapwise and edgewise blade modes. 

Using AeroDyn, ElastoDyn, and BeamDyn I've simulated a steady inflow case for a couple of seconds. From that simulation, I get the following loads and deflections: 

<img width="400" alt="loading" src="https://user-images.githubusercontent.com/39682562/225737483-309e9ee0-6b17-44c5-961f-450cf969597f.png"><img width="400" alt="loading" src="https://user-images.githubusercontent.com/39682562/225737628-e53a13a4-5efe-44e9-8856-059bb3aa4d39.png">

So the meat of my question is why are my Y deflections (B1TipTDyr) are negative? The X and Z deflections are as I expect. I have a positive Y loading (AB1NxxxFy), so I would expect a positive Y deflection. The loads across the radius of the blade are positive across time. There aren't any coupling terms in the stiffness matrix. So I thought that maybe there was some sort of rotation difference, but from what I've found in the documentation the reference frames should be more or less the same for my problem. 

<img width="600" alt="image" src="https://user-images.githubusercontent.com/39682562/225738666-8465f655-7e24-49a6-a31d-f36c53cde220.png">

<img width="450" alt="BeamDyn Sideview" src="https://user-images.githubusercontent.com/39682562/225738823-879cc5ca-b4de-421c-871d-23ab91f5e0ed.png"><img width="300" alt="BeamDyn Frontview" src="https://user-images.githubusercontent.com/39682562/225738757-3f5b70c3-3d80-49b1-bad3-cc17bbb1b438.png"> 

I feel like I'm missing something very obvious. Do you guys have any thoughts? I've attached my input files (save the airfoils).

Thanks in advance,

Adam

[sn5.zip](https://github.com/OpenFAST/openfast/files/10995423/sn5.zip)

**Comment by jjonkman** [comment]
Dear @Cardoza2,

Looking at your input files, it also looks like you disabled shaft tilt.  

Can you share plots of the aerodynamic load distribution (Fx, Fy, Mm) from BeamDyn or AeroDyn, as well as some measure of the elastic twist as function of radius at the last time step?

Best reagards,

**Reply by Cardoza2** [reply]
@jjonkman,

Thank you for taking a look at this. 
Yes, I set shaft tilt to zero. I attempted to simplify, and isolate the rotor with head on wind for a code to code comparison. There might be some other changes that I made and have forgotten about. Hopefully nothing critical. 

Here are the aerodynamic loadings for the final time step. Note, I didn't plot the moment, because I set `UseBlCm` to false and when I've done that in the past the reported aerodynamic moment was always zero. 
![image](https://user-images.githubusercontent.com/39682562/225792360-0d648173-d01d-4730-8fd1-7b461a61dd73.png)

Here is the Weiner-Milenkovic parameters for the blade at the final time step:
![image](https://user-images.githubusercontent.com/39682562/225792538-76726c60-311e-4b85-9b27-562744f91ad9.png)

Regards,

**Reply by jjonkman** [reply]
Thanks for sharing.  Can you clarify which distributed load outputs you are plotting?

Best regards,

**Reply by Cardoza2** [reply]
Sure thing, that's `AB1NxxxFx` and `AB1NxxxFy` from AeroDyn, so the the outputs "Fx, Fy" in the nodal outputs, plotted respective to their nodal radii. 

Best,

**Reply by jjonkman** [reply]
Dear @Cardoza2,

I'm a bit puzzled now too.  The deflections and rotations seem consistent with each other, but would imply a negative Fy force.  I'm not sure why a positive Fy force is shown.  I'm not aware of any problems with this part of the software.

Any chance your plotting script could be swapping the sign of Fy?

Can you also output the applied forces seen by BeamDyn, e.g., DFxr and DFyr?

Best regards,

**Reply by Cardoza2** [reply]
@jjonkman, 

I'm glad I'm not the only one. I just double checked and I don't swap the sign of Fy. I'm getting the same loads out of CCBlade.jl. 

I'm struggling to get DFxr, DFyr, and DFzr to output anything other than zeros. Additionally, the units for these values in the output file are listed as `INVALID`. 

Although, upon further consideration, I think I may have realized the rub here. The loads Fx and Fy are the dimensionalized forms of $C_n$ and $C_c$ respectively; and in my experience (albeit that that's not worth a lot yet) they are reported positive when they are in the directions shown below. Although, as also shown in the figure below, Fy is clearly in the negative Y direction... Which would explain why I'm getting negative deflections. I think that the axes shown below are the same as the axes that I showed before. Does that logic pan out? 

![image](https://user-images.githubusercontent.com/39682562/226027790-68a6fc6b-6c43-4c46-a3f7-14bd524bec07.png)

regards,

**Reply by jjonkman** [reply]
Dear @Cardoza2,

I was originally thinking that this morning, but when looking at the AeroDyn source code, I noticed that the sign of the Fy output gets swapped relative to the Cc direction (as expected), so, AeroDyn is really reporting the force along +y, not the force along -y.  And a force along +y should result in deflection along +y.

Regarding BeamDyn nodal outputs `DFxr` and `DFyr`, did you place those in the nodal output section of the BeamDyn input file?  Do other BeamDyn nodal outputs work such as `Fxr` and `Fyr`?

Best regards,

**Reply by Cardoza2** [reply]
Dear @jjonkman, 

Yeah, I think I see the same thing. Like the negative on line 3129 of AeroDyn.f90? 
```Fortran
Ct  = -Cl*sin(aoa) + Cd*cos(aoa)
```

Might it have to do with how the output is written to file? I see a negative on the write output on line 791 of AeroDyn_AllBldNdOuts_IO.f90. I don't know if that's what we're seeing. 

```Fortran
         CASE ( BldNd_Fy )
            DO IdxBlade=1,p%BldNd_BladesOut
               DO IdxNode=1,p%NumBlNds   
                  y%WriteOutput( OutIdx )  = -m%Y(IdxNode,IdxBlade)
                  OutIdx = OutIdx + 1
               END DO
            END DO 
```

I did put the `DFxr` and `DFyr` in the nodal outputs. The force resultants (`Fxr` and `Fyr`) give values in the nodal outputs, and their signs are consistent with the deflections: 
![image](https://user-images.githubusercontent.com/39682562/226084562-41e401b6-48d0-461a-a05f-951a7267f59d.png)

Best regards,

**Reply by jjonkman** [reply]
Dear @Cardoza2,

OK, it looks like BeamDyn is internally self-consistent, which is good.

It looks like I made a mistake on Friday.  I saw the code snipped you pasted regarding `BldNd_Fy` and thought that looked correct, because I thought m%Y was positive in the Cc direction.  However, there is actually a double negative that I didn't notice.  In line 3146 of AeroDyn.f90, AeroDyn uses the variable `m%BEMT_y%cy` for Cc, and there is a minus sign used in the equation for m%Y.  This means that AeroDyn output `Fy` is actually in the -Y direction, and now everything is consistent.

I am questioning, though, why AeroDyn outputs `Fy` in the -Y direction.  Perhaps this needs to be changed?

@ebranlard, @andrew-platt, @bjonkman -- do you have an opinion on this?

Best regards,
